tests/unit/test_charm_state.py

```diff
@@ -929,6 +929,41 @@ def mock_charm_state_data():
     }
 
 
+@pytest.mark.parametrize(
+    "immutable_config",
+    [
+        pytest.param("runner_storage", id="Runner storage"),
+        pytest.param("base_image", id="Base image"),
+    ],
+)
+def test_check_immutable_config_key_error(
+    mock_charm_state_path: Path,
+    mock_charm_state_data: dict[str, typing.Any],
+    immutable_config: str,
+    monkeypatch: pytest.MonkeyPatch,
+    caplog: pytest.LogCaptureFixture,
+):
+    """
+    arrange: Mock CHARM_STATE_PATH and read_text method to return previous state with same config.
+    act: Call _check_immutable_config_change method.
+    assert: None is returned.
+    """
+    mock_charm_state_data["runner_config"].pop(immutable_config)
```

Comment:
Shouldn't we use a cloned object, since you reuse it in line 958 (and I think you want to test that the key is included in current state but not in the previous)?

Reply:
Good point, i thought a function scoped fixture would regenerate the object, let me test it!

Reply:
My point is not that it gets regenerated, but that you change it within the test itself. So the question would be, should the state in line 958 contain the `immutable_config` or not?

Reply:
I see what you mean, good point, the method under test is a class method so we don't quite need to instantiate the charm state at all in the first place.
Updated code, thank you!